src/megatron/energon/wrappers/blend_dataset.py

```diff
@@ -77,18 +77,31 @@ def __len__(self) -> int:
 
     def __iter__(self) -> Iterator[T_sample]:
         assert self.worker_has_samples(), "Cannot blend all empty datasets"
-        datasets, weights = zip(
-            *[
-                (dataset, weight)
-                for dataset, weight in self.dataset_weights
-                if dataset.worker_has_samples()
-            ]
-        )
-        dataset_iters = [iter(dataset) for dataset in datasets]
-        weights = torch.tensor(weights, dtype=torch.float32)
-        probs = weights / weights.sum()
-        assert torch.all(probs > 0), "Negative weights are not allowed"
 
+        # Create a list of datasets and their weights, but
+        # set the weight to 0 if the dataset has no samples on this worker.
+
+        dataset_iters = []
+        weights = []
+        for idx, (dataset, weight) in enumerate(self.dataset_weights):
+            assert weight > 0, "All blending weights must be > 0"
+
+            if dataset.worker_has_samples():
+                dataset_iters.append(iter(dataset))
+                weights.append(weight)
+            else:
+                dataset_iters.append(None)
+                weights.append(0)
+
+        probs = torch.tensor(weights, dtype=torch.float32)
+        if probs.sum() == 0:
+            raise RuntimeError(
+                "There is a worker with no samples in any of the blended datasets. "
+                "This can happen if you have a lot of workers and many small datasets. "
+                "Currently this case is not supported."
```

Comment:
```suggestion
                "There is a worker with no samples in any of the blended datasets. "
                "This can happen if you have a lot of workers and your dataset is too small. "
                "Currently this case is not supported."
```
With rotation, the number of datasets does not play a role any more.